## Exploratory Data Analysis 

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import pickle
import functions as fn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

In [3]:
car_df = pd.read_csv('data/cardataset.zip')

In [4]:
car_df.shape

(11914, 16)

In [5]:
car_df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [6]:
car_df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

subsetting the dataset to 7920 rows from 11914 total rows

In [7]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
Make                 11914 non-null object
Model                11914 non-null object
Year                 11914 non-null int64
Engine Fuel Type     11911 non-null object
Engine HP            11845 non-null float64
Engine Cylinders     11884 non-null float64
Transmission Type    11914 non-null object
Driven_Wheels        11914 non-null object
Number of Doors      11908 non-null float64
Market Category      8172 non-null object
Vehicle Size         11914 non-null object
Vehicle Style        11914 non-null object
highway MPG          11914 non-null int64
city mpg             11914 non-null int64
Popularity           11914 non-null int64
MSRP                 11914 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 1.5+ MB


In [8]:
# plt.hist(car_df['Year']);
car_df1 = car_df.loc[car_df['Year'] > 2009]

In [9]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP              63
Engine Cylinders       20
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      1642
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [10]:
car_df1.dropna(subset=['Number of Doors'], axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [11]:
car_df1.dropna(subset=['Engine HP'], axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [12]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP               0
Engine Cylinders       19
Transmission Type       0
Driven_Wheels           0
Number of Doors         0
Market Category      1637
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [14]:
car_df1.loc[car_df1['Engine Cylinders'].isna()] = car_df1.loc[car_df1['Engine Cylinders'].isna()].fillna(0)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [15]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP               0
Engine Cylinders        0
Transmission Type       0
Driven_Wheels           0
Number of Doors         0
Market Category      1637
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [19]:
car_df1.loc[car_df1['Market Category'].isna()]['Make'].value_counts()

Toyota        328
Nissan        197
Chevrolet     174
Volkswagen    170
Honda         139
Ford          124
GMC            76
Suzuki         68
Hyundai        62
Kia            59
Dodge          56
Mazda          44
Mitsubishi     30
Subaru         30
Chrysler       29
FIAT           21
Buick          17
Scion          12
Pontiac         1
Name: Make, dtype: int64

In [20]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7856 entries, 0 to 11912
Data columns (total 16 columns):
Make                 7856 non-null object
Model                7856 non-null object
Year                 7856 non-null int64
Engine Fuel Type     7856 non-null object
Engine HP            7856 non-null float64
Engine Cylinders     7856 non-null float64
Transmission Type    7856 non-null object
Driven_Wheels        7856 non-null object
Number of Doors      7856 non-null float64
Market Category      6219 non-null object
Vehicle Size         7856 non-null object
Vehicle Style        7856 non-null object
highway MPG          7856 non-null int64
city mpg             7856 non-null int64
Popularity           7856 non-null int64
MSRP                 7856 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 1.0+ MB


In [21]:
car_df1.dropna(axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [22]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6219 entries, 0 to 11912
Data columns (total 16 columns):
Make                 6219 non-null object
Model                6219 non-null object
Year                 6219 non-null int64
Engine Fuel Type     6219 non-null object
Engine HP            6219 non-null float64
Engine Cylinders     6219 non-null float64
Transmission Type    6219 non-null object
Driven_Wheels        6219 non-null object
Number of Doors      6219 non-null float64
Market Category      6219 non-null object
Vehicle Size         6219 non-null object
Vehicle Style        6219 non-null object
highway MPG          6219 non-null int64
city mpg             6219 non-null int64
Popularity           6219 non-null int64
MSRP                 6219 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 826.0+ KB


In [23]:
car_df1.duplicated().sum()

241

In [24]:
car_df1.drop_duplicates(inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [25]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5978 entries, 0 to 11912
Data columns (total 16 columns):
Make                 5978 non-null object
Model                5978 non-null object
Year                 5978 non-null int64
Engine Fuel Type     5978 non-null object
Engine HP            5978 non-null float64
Engine Cylinders     5978 non-null float64
Transmission Type    5978 non-null object
Driven_Wheels        5978 non-null object
Number of Doors      5978 non-null float64
Market Category      5978 non-null object
Vehicle Size         5978 non-null object
Vehicle Style        5978 non-null object
highway MPG          5978 non-null int64
city mpg             5978 non-null int64
Popularity           5978 non-null int64
MSRP                 5978 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 794.0+ KB


In [28]:
# with open('car_df1.pickle', 'wb') as f:
#     pickle.dump(car_df1, f, pickle.HIGHEST_PROTOCOL)

create a new column to assign carmaker origin to each row 

In [29]:
with open('make_origin.pickle', 'rb') as f:
    make_origin_dict = pickle.load(f)

In [30]:
make_origin_dict

{'BMW': 'Germany',
 'FIAT': 'Italy',
 'Chrysler': 'USA',
 'Mazda': 'Japan',
 'Nissan': 'Japan',
 'Ferrari': 'Italy',
 'Alfa Romeo': 'Italy',
 'McLaren': 'UK',
 'Maybach': 'Germany',
 'Porsche': 'Germany',
 'Toyota': 'Japan',
 'Saab': 'Sweden',
 'Audi': 'Germany',
 'GMC': 'USA',
 'Hyundai': 'Korea',
 'Honda': 'Japan',
 'Mercedes-Benz': 'Germany',
 'Cadillac': 'USA',
 'Chevrolet': 'USA',
 'Dodge': 'USA',
 'Lamborghini': 'Italy',
 'Bentley': 'UK',
 'Volkswagen': 'Germany',
 'Subaru': 'Japan',
 'Ford': 'USA',
 'Volvo': 'Sweden',
 'Kia': 'Korea',
 'Lincoln': 'USA',
 'Lexus': 'Japan',
 'Rolls-Royce': 'UK',
 'Aston Martin': 'UK',
 'Land Rover': 'UK',
 'Mitsubishi': 'Japan',
 'Lotus': 'UK',
 'Buick': 'USA',
 'Infiniti': 'Japan',
 'Scion': 'Japan',
 'Genesis': 'Korea',
 'Maserati': 'Italy',
 'Suzuki': 'Japan',
 'HUMMER': 'USA',
 'Acura': 'Japan',
 'Pontiac': 'USA'}

split the attributes in the column market catergory to seperate columns

In [31]:
car_df1 = fn.market_columns(car_df1)

/Users/Erica/flatiron/mod_5/mod_5_project/car-classification-using-machine-learning/functions.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  df[col] = df['Market Category'].apply(lambda x: 1 if col in x.split(',') else 0)
/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [32]:
car_df1.drop(columns='Number of Doors', axis=1, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [33]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5978 entries, 0 to 11912
Data columns (total 24 columns):
Make                 5978 non-null object
Model                5978 non-null object
Year                 5978 non-null int64
Engine Fuel Type     5978 non-null object
Engine HP            5978 non-null float64
Engine Cylinders     5978 non-null float64
Transmission Type    5978 non-null object
Driven_Wheels        5978 non-null object
Vehicle Size         5978 non-null object
Vehicle Style        5978 non-null object
highway MPG          5978 non-null int64
city mpg             5978 non-null int64
Popularity           5978 non-null int64
MSRP                 5978 non-null int64
High-Performance     5978 non-null int64
Exotic               5978 non-null int64
Crossover            5978 non-null int64
Flex Fuel            5978 non-null int64
Factory Tuner        5978 non-null int64
Performance          5978 non-null int64
Luxury               5978 non-null int64
Hatchback           

In [34]:
car_df1['Vehicle Style'].value_counts()

4dr SUV                1816
Sedan                  1401
Coupe                   667
4dr Hatchback           557
Convertible             455
Crew Cab Pickup         260
Wagon                   225
2dr Hatchback           189
Extended Cab Pickup     119
Regular Cab Pickup      105
Passenger Van           100
Passenger Minivan        38
Cargo Van                34
2dr SUV                   6
Convertible SUV           5
Cargo Minivan             1
Name: Vehicle Style, dtype: int64

In [35]:
car_df1 = car_df1.loc[(car_df1['Vehicle Style'] != '2dr SUV') 
                      & (car_df1['Vehicle Style'] != 'Convertible SUV') &
         (car_df1['Vehicle Style'] != 'Cargo Minivan')]

In [36]:
car_df1['Driven_Wheels'].value_counts()

front wheel drive    1944
all wheel drive      1841
rear wheel drive     1679
four wheel drive      502
Name: Driven_Wheels, dtype: int64

In [37]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5966 entries, 0 to 11912
Data columns (total 24 columns):
Make                 5966 non-null object
Model                5966 non-null object
Year                 5966 non-null int64
Engine Fuel Type     5966 non-null object
Engine HP            5966 non-null float64
Engine Cylinders     5966 non-null float64
Transmission Type    5966 non-null object
Driven_Wheels        5966 non-null object
Vehicle Size         5966 non-null object
Vehicle Style        5966 non-null object
highway MPG          5966 non-null int64
city mpg             5966 non-null int64
Popularity           5966 non-null int64
MSRP                 5966 non-null int64
High-Performance     5966 non-null int64
Exotic               5966 non-null int64
Crossover            5966 non-null int64
Flex Fuel            5966 non-null int64
Factory Tuner        5966 non-null int64
Performance          5966 non-null int64
Luxury               5966 non-null int64
Hatchback           

In [38]:
car_df1['Origin'] = car_df1['Make'].apply(lambda m: make_origin_dict[m])

In [39]:
car_df1.drop(['Make', 'Model'], axis=1, inplace=True)

In [40]:
X = car_df1.drop('Origin', axis=1)
y = car_df1['Origin']

In [41]:
y

0        Germany
1        Germany
2        Germany
3        Germany
4        Germany
5        Germany
6        Germany
7        Germany
8        Germany
9        Germany
10       Germany
11       Germany
12       Germany
13       Germany
15       Germany
16       Germany
32         Italy
33         Italy
34         Italy
41       Germany
42       Germany
43       Germany
44       Germany
45       Germany
46       Germany
47       Germany
48       Germany
49       Germany
50       Germany
51       Germany
          ...   
11865        USA
11866        USA
11869        USA
11870        USA
11871        USA
11872        USA
11873        USA
11874        USA
11875        USA
11876        USA
11877        USA
11878        USA
11879        USA
11880        USA
11894    Germany
11895    Germany
11896    Germany
11897    Germany
11898    Germany
11899    Germany
11900    Germany
11901    Germany
11902    Germany
11906      Japan
11907      Japan
11908      Japan
11909      Japan
11910      Jap

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [43]:
# Create train and test dataframes with only the categorical variables

X_train_obj = X_train[[col for col,dtype in list(zip(X_train.columns, X_train.dtypes)) 
                          if dtype == np.dtype('O')]]
X_test_obj = X_test[[col for col,dtype in list(zip(X_test.columns, X_test.dtypes)) 
                        if dtype == np.dtype('O')]]

In [44]:
X_train_nonobj = X_train[[col for col,dtype in list(zip(X_train.columns, X_train.dtypes)) 
                          if dtype != np.dtype('O')]]
X_test_nonobj = X_test[[col for col,dtype in list(zip(X_test.columns, X_test.dtypes)) 
                        if dtype != np.dtype('O')]]

In [45]:
ohe = OneHotEncoder(handle_unknown='ignore')
X_train_obj_ohe = ohe.fit_transform(X_train_obj)
X_test_obj_ohe = ohe.transform(X_test_obj)

In [46]:
X_train_nonobj_df = pd.DataFrame(X_train_nonobj).reset_index(drop=True) 
X_train_obj_ohe_df = pd.DataFrame(X_train_obj_ohe.todense(), columns=ohe.get_feature_names()).reset_index(drop=True)

In [47]:
X_train_all = pd.concat([X_train_nonobj_df, X_train_obj_ohe_df], axis=1)

In [48]:
X_test_nonobj_df = pd.DataFrame(X_test_nonobj).reset_index(drop=True) 
X_test_obj_ohe_df = pd.DataFrame(X_test_obj_ohe.todense(), columns=ohe.get_feature_names()).reset_index(drop=True)

In [52]:
X_test_all = pd.concat([X_test_nonobj_df, X_test_obj_ohe_df], axis=1)

In [54]:
X_train_all.shape

(4474, 49)

In [55]:
X_test_all.shape

(1492, 49)

In [56]:
origins = list(set(make_origin_dict.values()))

origin_code = {}
for i in range(len(origins)):
    origin_code[origins[i]] = i

y_train = y_train.apply(lambda x: origin_code[x])
y_test = y_test.apply(lambda x: origin_code[x])

In [ ]:
# origins = list(set(make_origin_dict.values()))

# origin_code = {}
# for i in range(len(origins)):
#     origin_code[origins[i]] = i
# origin_code

# car_df1['origin_code'] = car_df1['Origin'].apply(lambda x: origin_code[x])
# car_df1.drop(columns='Origin', axis=1, inplace=True)

In [ ]:
# X_train_all = pd.concat([pd.DataFrame(X_train_nonobj), pd.DataFrame(X_train_obj_ohe.todense(), columns=ohe.get_feature_names())], axis=1)
# X_test_all = pd.concat([pd.DataFrame(X_test_nonobj), pd.DataFrame(X_test_obj_ohe.todense(), columns=ohe.get_feature_names())], axis=1)

In [ ]:
# cat_features = ['Engine Fuel Type', 'Transmission Type', 'Driven_Wheels', 'Vehicle Size', 'Vehicle Style']
# cat_index=[]
# for ind, val in enumerate(X_train.columns):
#     #print(ind, val)
#     if val in cat_features:
#         cat_index.append(ind)

In [ ]:
# X_train_df = pd.DataFrame(X_train_ohe.todense(), columns=ohe.get_feature_names())
# X_train_df.columns